File: android/app/src/main/java/com/riz/app/crypto/PasswordPolicy.kt
package com.riz.app.crypto

import android.content.Context
import com.nulabinc.zxcvbn.Zxcvbn
import com.riz.app.util.AppLog
import java.security.SecureRandom

/**
 * Password strength scoring + EFF Large Wordlist passphrase generator.
 */
object PasswordPolicy {
    const val MIN_SCORE = 3
    const val DEFAULT_PASSPHRASE_WORDS = 6
    private const val SEPARATOR = "-"
    private const val EXPECTED_WORDLIST_SIZE = 7776
    private const val WORDLIST_ASSET = "eff_large_wordlist.txt"

    /**
     * Cap input to zxcvbn at this length. Above 64 chars the analyser cost
     * grows steeply and any password that long is already past the
     * unguessable threshold.
     */
    private const val STRENGTH_INPUT_CAP = 64

    /** 2^32 — used as the upper bound for unbiased 32-bit rejection sampling. */
    private const val UINT32_RANGE = 0x1_0000_0000L

    /** Mask to interpret Int.nextInt() as an unsigned 32-bit Long. */
    private const val UINT32_MASK = 0xFFFF_FFFFL

    private val zxcvbn = Zxcvbn()
    private val secureRandom = SecureRandom()
    private var words: List<String> = emptyList()

    fun init(context: Context) {
        if (words.isNotEmpty()) return
        try {
            words =
                context.assets.open(WORDLIST_ASSET).bufferedReader().use { reader ->
                    reader.readLines().filter { it.isNotBlank() }
                }
        } catch (e: Exception) {
            AppLog.e("PasswordPolicy", "Failed to load wordlist", e)
        }
    }

    data class Assessment(
        val score: Int,
        val warning: String,
        val suggestions: List<String>,
    )

    fun assessStrength(password: String): Assessment {
        val sample =
            if (password.length > STRENGTH_INPUT_CAP) {
                password.substring(0, STRENGTH_INPUT_CAP)
            } else {
                password
            }
        val strength = zxcvbn.measure(sample)
        val feedback = strength.feedback
        val suggestions = feedback?.getSuggestions(java.util.Locale.ENGLISH).orEmpty()
        return Assessment(
            score = strength.score,
            warning = feedback?.getWarning(java.util.Locale.ENGLISH).orEmpty(),
            suggestions = suggestions,
        )
    }

    /**
     * Uniform-random passphrase via rejection sampling on 32-bit ints.
     */
    fun generatePassphrase(numWords: Int = DEFAULT_PASSPHRASE_WORDS): String = pickRandomWords(numWords).joinToString(SEPARATOR)

    fun pickRandomWords(numWords: Int): List<String> {
        check(words.size >= EXPECTED_WORDLIST_SIZE) {
            "Wordlist not loaded; call PasswordPolicy.init(context) first"
        }
        val n = words.size
        // Largest multiple of n that fits in an unsigned 32-bit int.
        val limit = (UINT32_RANGE / n) * n
        return List(numWords) {
            var r: Long
            do {
                r = secureRandom.nextInt().toLong() and UINT32_MASK
            } while (r >= limit)
            words[(r % n).toInt()]
        }
    }
}

File: src/core/passwordPolicy.js
/**
 * Password policy: strength scoring + EFF Large Wordlist passphrase generator.
 */
import { zxcvbn, zxcvbnOptions } from '@zxcvbn-ts/core';
import * as zxcvbnCommonPackage from '@zxcvbn-ts/language-common';
import { EFF_LARGE_WORDLIST } from './wordlist.js';

zxcvbnOptions.setOptions({
  graphs: zxcvbnCommonPackage.adjacencyGraphs,
  dictionary: { ...zxcvbnCommonPackage.dictionary },
});

const DICEWARE = EFF_LARGE_WORDLIST;

export const MIN_SCORE = 3;
export const DEFAULT_PASSPHRASE_WORDS = 6;
const SEPARATOR = '-';

/**
 * Returns { score (0..4), warning, suggestions, crackTimesDisplay, guessesLog10 }.
 */
export function assessStrength(pwd) {
  const result = zxcvbn(pwd || '');
  return {
    score: result.score,
    warning: result.feedback?.warning || '',
    suggestions: result.feedback?.suggestions || [],
    crackTimesDisplay: result.crackTimesDisplay,
    guessesLog10: result.guessesLog10,
  };
}

/**
 * Generates a passphrase using uniform rejection sampling over the EFF list.
 * Each draw uses 32 bits from crypto.getRandomValues; values outside the
 * largest multiple of N are rejected to avoid modulo bias.
 */
export function generatePassphrase(numWords = DEFAULT_PASSPHRASE_WORDS) {
  if (!DICEWARE || DICEWARE.length < 7776) {
    throw new Error('Diceware wordlist unavailable');
  }
  const N = DICEWARE.length;
  const limit = Math.floor(0x100000000 / N) * N;
  const buf = new Uint32Array(1);
  const out = new Array(numWords);
  for (let i = 0; i < numWords; i++) {
    let r;
    do {
      crypto.getRandomValues(buf);
      r = buf[0];
    } while (r >= limit);
    out[i] = DICEWARE[r % N];
  }
  return out.join(SEPARATOR);
}
